include/ddsrouter/communication/payload_pool/PayloadPool.hpp

```diff
@@ -0,0 +1,246 @@
+     *
+     * @return true if everything ok
+     * @return false if something went wrong
+     *
+     * @pre Fields @c payload must not have been initialized.
+     */
+    virtual bool get_payload(
+            uint32_t size,
+            Payload& payload) = 0;
+
+    /**
+     * @brief Store in \c target_payload the data from \c src_payload .
+     *
+     * This method set \c target_payload fields \c max_size , \c lenght and \c data .
+     * This method "should" reuse data in \c src_payload and not copy it in case \c data_owner is \c this .
+     *
+     * @note This method may reserve new memory in case the owner is not \c this .
+     *
+     * @param [in] src_payload     Payload to move to target
+     * @param [in,out] data_owner      Payload pool owning incoming data \c src_payload
+     * @param [out] target_payload  Payload to assign the payload to
+     *
+     * @warning @c data_owner can only be changed from @c nullptr to @c this. If a value different from
+     * @c nullptr is received it must be left unchanged.
+     *
+     * @return true if everything ok
+     * @return false if something went wrong
+     *
+     * @pre Fields @c target_payload must not have been initialized.
+     */
+    virtual bool get_payload(
+            const Payload& src_payload,
+            IPayloadPool*& data_owner,
+            Payload& target_payload) = 0;
+
+    /**
+     * @brief Release the data from the \c payload .
+     *
+     * @note This method must only release the actual memory of a data in case nobody is referencing it anymore.
+     *
+     * @note This method should use method \c reserve_ to reserve new memory.
+     *
+     * Reset the \c payload info.
+     *
+     * @param [in,out] payload Payload to release data from
+     *
+     * @return true if everything ok
+     * @return false if something went wrong
+     *
+     * @pre @c payload must have been initialized from this pool.
+     */
+    virtual bool release_payload(
+            Payload& payload) = 0;
+
+    //! Wether every payload get has been released.
+    virtual bool is_clean() const noexcept;
+
+protected:
+
+    /**
+     * @brief Reserve a new space of memory for new data.
+     *
+     * It increases \c reserve_count_ .
+     *
+     * @param size size of memory chunk to reserve
+     * @param payload object where introduce the new data pointer
+     *
+     * @return true if everything ok
+     * @return false if something went wrong
+     */
+    virtual bool reserve_(
+            uint32_t size,
+            Payload& payload);
+
+    /**
+     * @brief Free a memory space.
+     *
+     * It increases \c release_count_ .
+     *
+     * @param payload object to free the data from
+     *
+     * @return true if everything ok
+     * @return false if something went wrong
+     *
+     * @throw \c IncosistencyException if more releases than reserves has been done
+     */
+    virtual bool release_(
```

Comment:
Check that the `payload.data != nullptr ` to validate the output of the `empty()` function